mava/core_jax_test.py

```diff
@@ -0,0 +1,88 @@
+# python3
+# Copyright 2021 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Tests for core Mava interfaces for Jax systems."""
+
+from typing import Any, List
+
+import pytest
+
+from mava.core_jax import BaseSystem
+
+# Dummy component
+COMPONENT = None
+
+
+class TestDummySystem(BaseSystem):
+    """Create a complete class with all abstract method overwritten."""
+
+    def design(self) -> List[Any]:
+        """Dummy design"""
+        self.components = [COMPONENT, COMPONENT]
+        assert len(self.components) == 2
+        return self.components
+
+    def update(self, component: Any, name: str) -> None:
+        """Dummy update"""
+        assert component is None
+        assert name == "update"
+
+    def add(self, component: Any, name: str) -> None:
+        """Dummy add"""
+        assert component is None
+        assert name == "add"
+
+    def configure(self, **kwargs: Any) -> None:
+        """Dummy configure"""
+        assert kwargs["param_0"] == 0
+        assert kwargs["param_1"] == 1
+
+    def launch(
+        self,
+        num_executors: int,
+        nodes_on_gpu: List[str],
+        multi_process: bool = True,
+        name: str = "system",
+    ) -> None:
+        """Dummy launch"""
+        assert num_executors == 1
+        assert multi_process is True
+        assert nodes_on_gpu[0] == "process"
+        assert name == "system"
+
+
+@pytest.fixture
+def dummy_system() -> TestDummySystem:
+    """Create complete system for use in tests"""
+    return TestDummySystem()
+
+
+def test_dummy_system(dummy_system: TestDummySystem) -> None:
```

Comment:
@arnupretorius I am not too sure on the utility of these tests. If we are testing the abstract base class, then we should tests things like no subclass can implement it without implement those abstract methods or test something more foundational. 

Here it looks like we are mocking out a system and then testing the mocked system, without testing any real functionality. I think other things we could test, could be for example , have `BaseSystem` class create properties or variables according to inputs and test those exist e.g. if we pass components to a subclass of `BaseSystem` , then those components should be returned when we call `design()` or actually implement `update()` or `add()`. 

My main worry here is that it doesn't seem to be testing any real code.

Reply:
I agree, the tests aren't really doing much here. Initially, all I wanted to test was what you suggested, that no subclass can implement it without implementing the abstract methods. However, I ran into the mypy errors I shared with you a while ago, where even if I use pytest's with raises exception or a try catch to have the test pass when the class has not implemented the required functions, mypy complains. So I simply remove that test. As for the other tests you are suggesting, I think they are more appropriate for when we are actually implementing the general class that will be shared between systems, as was done in this PR: feature/jax-general-system-class

Reply:
If you can find a way around the mypy issue, that would really help. 😉  The `with pytest.raises()` works nicely in some of the other tests, but for some reason doesn't want to play nice with mypy regarding this specific test of inheritance from an abstract class.

Reply:
Will look into it :)

Reply:
Hey @KaleabTessera, I think I found a workaround. Please see the updated tests. :)

Reply:
Thanks @arnupretorius , I really like the `test_exception_for_incomplete_child_class` test :open_hands: 

Personally, I still recommend not adding `TestDummySystem` and `test_dummy_system` until we are testing some implemented functionality. I think this PR is good with the abstract class + `test_exception_for_incomplete_child_class`. In future, we can implement tests for things like component adds etc.

Reply:
Cool, I removed the dummy system test. 👍